hypervisor/xenpv/xenpv.go

```diff
@@ -0,0 +1,285 @@
+	driver *XenPvDriver
+	domId  xl.Domid
+}
+
+func InitDriver() *XenPvDriver {
+	cmd, err := exec.LookPath(XENLIGHT_EXEC)
+	if err != nil {
+		return nil
+	}
+
+	ctx := &xl.Context{}
+	ctx.Open()
+
+	ctx.SigChildHandle()
+
+	return &XenPvDriver{
+		executable: cmd,
+		ctx:        ctx,
+	}
+}
+
+func (xd *XenPvDriver) Name() string {
+	return "xenpv"
+}
+
+func (xd *XenPvDriver) InitContext(homeDir string) hypervisor.DriverContext {
+	return &XenPvContext{
+		driver: xd,
+	}
+}
+
+func (xd *XenPvDriver) LoadContext(persisted map[string]interface{}) (hypervisor.DriverContext, error) {
+	return &XenPvContext{
+		driver: xd,
+	}, nil
+}
+
+func (xd *XenPvDriver) SupportLazyMode() bool {
+	return false
+}
+
+func (xd *XenPvDriver) SupportVmSocket() bool {
+	return false
+}
+
+func (xc *XenPvContext) Launch(ctx *hypervisor.VmContext) {
+	if xc.driver.executable == "" {
+		glog.Errorf("can not find xl executable")
+		ctx.Hub <- &hypervisor.VmStartFailEvent{Message: "can not find xl executable"}
+		return
+	}
+	uuid, err := xl.GenerateUuid()
+	if err != nil {
+		glog.Errorf("generate uuid failed: %v\n", err)
+		ctx.Hub <- &hypervisor.VmStartFailEvent{Message: err.Error()}
+		return
+	}
+
+	boot := ctx.Boot
+	config := &xl.DomainConfig{
+		Cinfo: xl.CreateInfo{
+			Type: "pv",
+			Name: ctx.Id,
+			Uuid: uuid,
+		},
+		Binfo: xl.BuildInfo{
+			Kernel:            boot.Kernel,
+			Initrd:            boot.Initrd,
+			Cmdline:           "console=hvc0 " + api.HYPER_P9_USE_XEN,
+			ClaimMode:         "True",
+			MaxVcpus:          boot.CPU,
+			MaxMemory:         uint64(boot.Memory * 1024),
+			TargetMemory:      uint64(boot.Memory * 1024),
+			DeviceModeVersion: "qemu_xen",
+			PvInfo:            xl.PvInfo{},
+		},
+		Channels: []xl.ChannelInfo{
+			{
+				DevId:  0,
+				Name:   "sh.hyper.channel.0",
+				Socket: xl.SocketInfo{Path: ctx.HyperSockName},
+			},
+			{
+				DevId:  1,
+				Name:   "sh.hyper.channel.1",
+				Socket: xl.SocketInfo{Path: ctx.TtySockName},
+			},
+		},
+		P9: []xl.P9Info{
+			{
+				ShareTag:      hypervisor.ShareDirTag,
+				ShareDir:      ctx.ShareDir,
+				SecurityModel: "none",
+			},
+		},
+	}
+
+	b, err := json.Marshal(config)
+	if err != nil {
+		glog.Errorf("fail to marshal xen domain config: %v, %v", b, err)
+		ctx.Hub <- &hypervisor.VmStartFailEvent{Message: err.Error()}
+		return
+	}
+
+	xctx := xc.driver.ctx
+	domid, err := xctx.CreateNewDomainFromJson(string(b))
+	if err != nil {
+		glog.Errorf("fail to create xen pv domain: %v", err)
+		xctx.DestroyDomainByName(ctx.Id)
+		ctx.Hub <- &hypervisor.VmStartFailEvent{Message: err.Error()}
+		return
+	}
+
+	glog.Infof("create success, domid: %v\n", domid)
+
+	xc.domId = xl.Domid(domid)
+
+	xc.connectConsole(ctx.Id, xc.domId)
+	go func() {
+		//unpause dom until runv connected to the socket, otherwise may loss the ready message
+		ctx.WaitSockConnected()
+		xctx.DomainUnpause(xc.domId)
+	}()
+}
+
+func (xc *XenPvContext) connectConsole(vmid string, domId xl.Domid) error {
```

Comment:
I can only add another unix socket to pass console data, seems like pv doesn't have direct way to get console from unix socket.